### PR TITLE
Support Chain Parameters V2

### PR DESCRIPTION
**Pull request by EmilLa1**
## Purpose

Addresses https://github.com/Concordium/concordium-node/issues/638 by incorporating the new structure of chain parameters into the node.

This PR offers support for the new way of structuring chain parameters and also supports for the new chain parameters (V2) which is to be used in the new consensus protocol.  https://github.com/Concordium/concordium-base/pull/288

The new consensus protocol will be introduced in a later PR and so will the protocol update.

## Changes

The changes are mainly refactorings in order to accommodate the new structures for chain parameters and versioning hereof. 

But in particular the following has been added apart from the necessary refactorings.

- Handling of new chain update queues for Timeout parameters, Minimum block time parameter and block energy limit parameter. 
- Queries: `getBlockPendingUpdates` now supports the new chain parameters.
- Skov is now constrained by `IsConsensusV0 pv`

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-consensus/src/Concordium/GlobalState/Basic/BlockState/Updates.hs [hunk 1]
```diff
@@ -167,18 +170,45 @@ processUpdateQueues t (theUpdates, ars, ips) =
...
+                in  ( ConsensusParametersV1 newTimeoutParameters newMinBlockTime newBlockEnergyLimit,
+                      NoParam,
+                      newTOQueue,
+                      newMBTQueue,
+                      newBELQueue,
+                      Map.empty,
+                      resTimeoutParameters',
+                      resMinBlockTime',
+                      resBlockEnergyLimit'
+                    )
```
Comment (EmilLa1) [inline concordium-consensus/src/Concordium/GlobalState/Basic/BlockState/Updates.hs:205]: Note to self. Make this more readable.

File: concordium-consensus/src/Concordium/GlobalState/BlockState.hs [hunk 2]
```diff
@@ -857,19 +864,19 @@ class (BlockStateQuery m) => BlockStateOperations m where
...
+        (AVSupportsDelegation (AccountVersionFor (MPV m))) =>
         UpdatableBlockState m ->
         m ([ActiveBakerInfo m], [ActiveDelegatorInfo])
 
     -- |Get the bakers for the epoch in which the block was baked.
     bsoGetCurrentEpochBakers :: UpdatableBlockState m -> m FullBakers
 
     -- |Get the bakers for the epoch in which the block was baked, together with their commission rates.
-    bsoGetCurrentEpochFullBakersEx :: (SupportsDelegation (MPV m)) => UpdatableBlockState m -> m FullBakersEx
+    bsoGetCurrentEpochFullBakersEx :: (AVSupportsDelegation (AccountVersionFor (MPV m))) => UpdatableBlockState m -> m FullBakersEx
```
Comment (EmilLa1) [inline concordium-consensus/src/Concordium/GlobalState/BlockState.hs:875]: Maybe we should introduce a type `SupportsDelegationPV = AVSupportsDelegation (AccountVersionFor pv)` in `base`

File: concordium-consensus/src/Concordium/GlobalState/BakerInfo.hs [hunk 3]
```diff
@@ -309,7 +309,7 @@ genesisBakerInfoEx spv cp GenesisBaker{..} = case spv of
               _bakerElectionVerifyKey = gbElectionVerifyKey,
               _bakerAggregationVerifyKey = gbAggregationVerifyKey
             }
-    binfoV1 :: (SupportsDelegation pv, ChainParametersVersionFor pv ~ 'ChainParametersV1) => BakerInfoEx (AccountVersionFor pv)
+    binfoV1 :: (AVSupportsDelegation (AccountVersionFor pv), PoolParametersVersionFor (ChainParametersVersionFor pv) ~ 'PoolParametersVersion1) => BakerInfoEx (AccountVersionFor pv)
```
Comment (EmilLa1) [inline concordium-consensus/src/Concordium/GlobalState/BakerInfo.hs:312]: Maybe we should use `SupportsDelegationPV` in these kind of places.

**Review by td202 — APPROVED** [review]

**Review by limemloh — APPROVED** [review]

File: concordium-consensus/src/Concordium/MultiVersion.hs [hunk 4]
```diff
@@ -403,42 +404,44 @@ newGenesis ::
...
-                notifyRegenesis (Just (genesisBlockHash gd))
+                            mvLog
+                    vcState <- newIORef st
+                    let vcShutdown = Skov.shutdownSkov vcContext =<< liftIO (readIORef vcState)
+                    let newEConfig :: VersionedConfiguration gsconf finconf pv
+                        newEConfig = VersionedConfiguration{..}
+                    writeIORef mvVersions (oldVersions `Vec.snoc` newVersion newEConfig)
+                    -- Notify the network layer we have a new genesis.
+                    notifyRegenesis (Just (genesisBlockHash gd))
+    SConsensusParametersVersion1 -> error "New consensus version not implemented yet." -- TODO: implement
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/MultiVersion.hs:444]: Please make a task that points to this. FIXME's and TODOs in the code should not be there without accompanying tasks.

File: concordium-consensus/src/Concordium/MultiVersion.hs [hunk 5]
```diff
@@ -466,53 +470,55 @@ checkForProtocolUpdate = liftSkov body
...
+                            vcState <- liftIO $ newIORef st
+                            let vcShutdown = Skov.shutdownSkov vcContext =<< liftIO (readIORef vcState)
+                            let newEConfig :: VersionedConfiguration gc fc newpv
+                                newEConfig = VersionedConfiguration{..}
+                            writeIORef mvVersions (existingVersions `Vec.snoc` newVersion newEConfig)
+                            -- Notify the network layer we have a new genesis.
+                            let Callbacks{..} = mvCallbacks
+                            liftIO $ notifyRegenesis (Just (regenesisBlockHash nextGenesis))
+                            return ()
+                    SConsensusParametersVersion1 -> error "New consensus version not implemented yet." -- TODO: implement
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/MultiVersion.hs:521]: Same here.

File: concordium-consensus/src/Concordium/MultiVersion.hs [hunk 6]
```diff
@@ -658,80 +664,82 @@ startupSkov genesis = do
...
-                -- last one we loaded.
-                Right (Just config@(EVersionedConfiguration newEConfig')) -> do
-                    mvrLogIO $ activateConfiguration config
-                    liftSkovUpdate newEConfig' checkForProtocolUpdate
+                    -- We loaded some protocol versions. Attempt to start in the
+                    -- last one we loaded.
+                    Right (Just config@(EVersionedConfiguration newEConfig')) -> do
+                        mvrLogIO $ activateConfiguration config
+                        liftSkovUpdate newEConfig' checkForProtocolUpdate
+            SConsensusParametersVersion1 -> error "New consensus not implemented yet." -- TODO: implement
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/MultiVersion.hs:742]: And here.

File: concordium-consensus/src/Concordium/MultiVersion.hs [hunk 7]
```diff
@@ -837,7 +845,7 @@ shutdownMultiVersionRunner MultiVersionRunner{..} = mask_ $ do
     -- Acquire the write lock. This prevents further updates, as they will block.
     takeMVar mvWriteLock
     versions <- readIORef mvVersions
-    runLoggerT (forM_ versions $ \(EVersionedConfiguration vc) -> vcShutdown vc) mvLog
+    runLoggerT (forM_ @_ @_ @_ @() versions $ \(EVersionedConfiguration vc) -> vcShutdown vc) mvLog
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/MultiVersion.hs:848]: It's entirely unclear what type parameters these _ are specifying unless you find the order in an obscure place  in the GHC manual. Please provide all of them.

**Comment by abizjak** [comment]
Before this can be merged at the following should be done
- merge with main to remove unused files
- Add tasks (or one task with a list) for all the remaining "TODOs" that are not implemented. It's hard to find random fixmes in a large codebase, so it is important we have them recorded. I understand that these would be easy to discover in testing as well.

File: concordium-consensus/src/Concordium/GlobalState/Basic/BlockState/Updates.hs [hunk 8]
```diff
@@ -167,18 +170,45 @@ processUpdateQueues t (theUpdates, ars, ips) =
     (updatedIPs, newAddIdentityProviderQueue, resAddIdentityProvider) = processARsAndIPsUpdates (IPS.idProviders ips) IPS.ipIdentity t _pAddIdentityProviderQueue
     (newTimeParameters, newTimeParametersQueue, resTimeParameters) =
         case chainParametersVersion @cpv of
-            SCPV0 -> (_cpTimeParameters, NothingForCPV1, Map.empty)
-            SCPV1 -> processValueUpdatesForCPV1 t _cpTimeParameters _pTimeParametersQueue & _3 %~ fmap UVTimeParameters
+            SChainParametersV0 -> (_cpTimeParameters, NoParam, Map.empty)
+            SChainParametersV1 -> processValueUpdatesO t (unOParam _cpTimeParameters) _pTimeParametersQueue & _3 %~ fmap UVTimeParameters & _1 %~ SomeParam
+            SChainParametersV2 -> processValueUpdatesO t (unOParam _cpTimeParameters) _pTimeParametersQueue & _3 %~ fmap UVTimeParameters & _1 %~ SomeParam
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/GlobalState/Basic/BlockState/Updates.hs]: Merged with main, so no longer a problem.

File: concordium-consensus/package.yaml [hunk 9]
```diff
@@ -58,6 +58,9 @@ dependencies:
 - lmdb
 - primitive
 - proto-lens >= 0.7
+- proto-lens-protobuf-types >= 0.7
+- proto-lens-runtime >= 0.7
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-consensus/package.yaml:62]: I don't think they are hmm.

File: concordium-consensus/src/Concordium/Scheduler/TreeStateEnvironment.hs [hunk 10]
```diff
@@ -953,7 +954,7 @@ mintAndReward bshandle blockParent slotNumber bid newEpoch mintParams mfinInfo t
         SP3 -> mintAndRewardCPV0AccountV0
         SP4 -> mintAndRewardCPV1AccountV1
         SP5 -> mintAndRewardCPV1AccountV1
-        SP6 -> mintAndRewardCPV1AccountV1
+        SP6 -> error "Minting undefined for P6" -- FIXME: implement
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/Scheduler/TreeStateEnvironment.hs:957]: Added a task: https://github.com/Concordium/concordium-node/issues/699

File: concordium-consensus/src/Concordium/Scheduler/TreeStateEnvironment.hs [hunk 11]
```diff
@@ -1053,7 +1054,7 @@ updateBirkParameters newSeedState bs0 oldChainParameters updates = case protocol
     SP3 -> updateCPV0AccountV0
     SP4 -> updateCPV1AccountV1
     SP5 -> updateCPV1AccountV1
-    SP6 -> updateCPV1AccountV1
+    SP6 -> error "updateBirkParameters not implemented for P6" -- FIXME: implement
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/Scheduler/TreeStateEnvironment.hs:1057]: Added this task for it: https://github.com/Concordium/concordium-node/issues/699

File: concordium-consensus/src/Concordium/MultiVersion.hs (hunk 4 above)
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/MultiVersion.hs:444]: Added this task. https://github.com/Concordium/concordium-node/issues/700

File: concordium-consensus/src/Concordium/MultiVersion.hs (hunk 5 above)
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/MultiVersion.hs:521]: Added this task: https://github.com/Concordium/concordium-node/issues/700

File: concordium-consensus/src/Concordium/MultiVersion.hs (hunk 6 above)
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/MultiVersion.hs:742]: Created this umbrella task for the todos/fixmes in the MultiVersionRuner. https://github.com/Concordium/concordium-node/issues/700

File: concordium-consensus/src/Concordium/Queries.hs [hunk 12]
```diff
@@ -261,7 +263,7 @@ liftSkovQueryBlockAndVersion a bh = MVR $ \mvr ->
 getConsensusStatus :: MVR gsconf finconf ConsensusStatus
 getConsensusStatus = MVR $ \mvr -> do
     versions <- readIORef (mvVersions mvr)
-    (csGenesisBlock, csGenesisTime) <- liftSkovQuery mvr (Vec.head versions) $ do
+    (csGenesisBlock, csGenesisTime) <- liftSkovQuery @_ @_ @(BlockHash, UTCTime) mvr (Vec.head versions) $ do
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/Queries.hs:266]: With the refactorings and removal of the basic block state these are not necessary anymore.

File: concordium-consensus/src/Concordium/MultiVersion.hs (hunk 7 above)
Comment (EmilLa1) [reply concordium-consensus/src/Concordium/MultiVersion.hs:848]: With the refactorings and removal of the basic block state these are not necessary anymore.

**Review by abizjak — APPROVED** [review]

**Pull request merged**